addons/pkg-npm/model-java/src/main/java/org/commonjava/indy/pkg/npm/model/PackageMetadata.java

```diff
@@ -261,9 +277,293 @@ public void setAttachments( Object attachments )
+                }
+                catch ( ParseException e )
+                {
+                    logger.error( String.format( "Cannot parse date when comparing. Reason: %s", e ) );
+                }
+                return compare;
+            } );
+
+            Map<String, String> clone = new LinkedHashMap<>();
+
+            if ( time.get( MODIFIED ) != null )
+            {
+                clone.put( MODIFIED, time.get( MODIFIED ) );
+            }
+            if ( time.get( CREATED ) != null )
+            {
+                clone.put( CREATED, time.get( CREATED ) );
+            }
+
+            for ( final Map.Entry<String, String> entry : timeList )
+            {
+                if ( MODIFIED.equals( entry.getKey() ) || CREATED.equals( entry.getKey() ) )
+                {
+                    continue;
+                }
+                clone.put( entry.getKey(), entry.getValue() );
+            }
+            time = clone;
+        }
+
+        // merge dist-tag object
+        DistTag sourceDist = source.getDistTags();
+        Map<String, String> sourceDistMap = sourceDist.fetchTagsMap();
+        Map<String, String> thisDistMap = distTags.fetchTagsMap();
+
+        if ( sourceDistMap.size() > 0 )
+        {
+            if ( thisDistMap == null || thisDistMap.size() <= 0 )
+            {
+                this.setDistTags( sourceDist );
+                changed = true;
+            }
+            else
+            {
+                for ( final String tag : sourceDistMap.keySet() )
+                {
+                    if ( !thisDistMap.keySet().contains( tag ) )
+                    {
+                        distTags.putTag( tag, sourceDistMap.get( tag ) );
+                        changed = true;
+                    }
+                    else
+                    {
+                        int compare = VersionUtils.createSingleVersion( sourceDistMap.get( tag ) )
+                                                  .compareTo( VersionUtils.createSingleVersion(
+                                                                  thisDistMap.get( tag ) ) );
+                        if ( compare > 0 )
+                        {
+                            distTags.putTag( tag, sourceDistMap.get( tag ) );
+                            changed = true;
+                        }
+                    }
+                }
+            }
+        }
+
+        //merge versions
+        Map<String, VersionMetadata> sourceVersions = source.getVersions();
+        Iterator vm = sourceVersions.keySet().iterator();
+
+        while ( vm.hasNext() )
+        {
+            String v = (String) vm.next();
+            VersionMetadata value = sourceVersions.get( v );
+            boolean s = false;
+            Iterator snapshot = versions.keySet().iterator();
+
+            while ( snapshot.hasNext() )
+            {
+                String preExisting = (String) snapshot.next();
+                if ( preExisting.equals( v ) )
```

Comment:
@yma88 I'd say it depends on how NPM works. For Maven, versions are supposed to be immutable once they're published in public. So if we come across one, we should keep it and avoid overwriting from another source...the same way we will prefer the content we get from the first source for a given path.

If NPM allows updating the same version (re-publishing it), then we may also need to think about what this means for retrieving the content itself...currently, Indy expects to use the first repo in a group that contains the requested path, in order to serve the content. If NPM allows re-publishing, it seems like we might have to track where the latest copy of the version metadata comes from and serve content from there instead of the first matching repo. This would be a major design change in Indy.

I'd say until we know this is something NPM / Node needs to support, we should keep the first version of metadata for a given version that we come across.

...Unless I'm misunderstanding what the version metadata actually means. Does it actually reference a series of publishing events, including pre-release and release pushes?